File: internal/manager/apikeyusage/parse.go
package apikeyusage

import (
	"bytes"
	"encoding/json"
	"io"
	"strconv"
	"strings"
	"time"
	"unicode"
	"unicode/utf8"

	"github.com/codeasier/mtls-router/internal/manager/protocol"
)

const (
	maxModels  = 64
	maxIDBytes = 256
)

// Period is the requested usage window.
type Period string

const (
	PeriodToday Period = "today"
	Period7d    Period = "7d"
	Period30d   Period = "30d"
)

// QuotaUnit is the optional remaining-budget unit.
type QuotaUnit string

const (
	QuotaUSD      QuotaUnit = "usd"
	QuotaTokens   QuotaUnit = "tokens"
	QuotaRequests QuotaUnit = "requests"
)

// Snapshot is one per-key usage document after fail-closed validation.
type Snapshot struct {
	Period  Period
	AsOf    string
	Summary Summary
	Quota   *Quota
	ByModel []Model
}

// Summary is the bounded numeric totals for one period.
type Summary struct {
	Requests         int64
	PromptTokens     int64
	CompletionTokens int64
	Cost             float64
}

// Quota is an optional remaining-budget projection for the same key.
type Quota struct {
	Used     float64
	Limit    *float64
	Unit     QuotaUnit
	ResetsAt string
}

// Model is one per-model row belonging to the authenticated key.
type Model struct {
	Model            string
	Requests         int64
	PromptTokens     int64
	CompletionTokens int64
	Cost             float64
}

// NormalizePeriod accepts the protocol period or the default 7d window.
func NormalizePeriod(value string) (Period, error) {
	switch strings.TrimSpace(value) {
	case "", string(Period7d):
		return Period7d, nil
	case string(PeriodToday):
		return PeriodToday, nil
	case string(Period30d):
		return Period30d, nil
	default:
		return "", &Error{Code: protocol.CodeInvalidParams, msg: "usage period is invalid"}
	}
}

// Parse validates one bounded /v1/usage body for the requested period.
func Parse(body []byte, period Period) (Snapshot, error) {
	if !utf8.Valid(body) {
		return Snapshot{}, responseInvalid()
	}
	decoder := json.NewDecoder(bytes.NewReader(body))
	decoder.UseNumber()
	token, err := decoder.Token()
	if err != nil || token != json.Delim('{') {
		return Snapshot{}, responseInvalid()
	}
	var snapshot Snapshot
	seen := map[string]bool{}
	for decoder.More() {
		keyToken, err := decoder.Token()
		key, ok := keyToken.(string)
		if err != nil || !ok || !seenField(seen, key) {
			return Snapshot{}, responseInvalid()
		}
		switch key {
		case "period":
			value, err := decodeString(decoder)
			if err != nil || Period(value) != period {
				return Snapshot{}, responseInvalid()
			}
			snapshot.Period = period
		case "as_of":
			value, err := decodeString(decoder)
			if err != nil || !validTime(value) {
				return Snapshot{}, responseInvalid()
			}
			snapshot.AsOf = value
		case "summary":
			summary, err := parseSummary(decoder)
			if err != nil {
				return Snapshot{}, err
			}
			snapshot.Summary = summary
		case "quota":
			quota, err := parseQuota(decoder)
			if err != nil {
				return Snapshot{}, err
			}
			snapshot.Quota = quota
		case "by_model":
			models, err := parseModels(decoder)
			if err != nil {
				return Snapshot{}, err
			}
			snapshot.ByModel = models
		default:
			if err := skipValue(decoder); err != nil {
				return Snapshot{}, responseInvalid()
			}
		}
	}
	if _, err := decoder.Token(); err != nil || !seen["period"] || !seen["summary"] {
		return Snapshot{}, responseInvalid()
	}
	if token, err := decoder.Token(); err != io.EOF || token != nil {
		return Snapshot{}, responseInvalid()
	}
	if snapshot.ByModel == nil {
		snapshot.ByModel = []Model{}
	}
	return snapshot, nil
}

func parseSummary(decoder *json.Decoder) (Summary, error) {
	fields, err := decodeObject(decoder)
	if err != nil {
		return Summary{}, err
	}
	requests, err := requiredCount(fields, "requests")
	if err != nil {
		return Summary{}, err
	}
	prompt, err := requiredCount(fields, "prompt_tokens")
	if err != nil {
		return Summary{}, err
	}
	completion, err := requiredCount(fields, "completion_tokens")
	if err != nil {
		return Summary{}, err
	}
	cost, err := requiredAmount(fields, "cost")
	if err != nil {
		return Summary{}, err
	}
	return Summary{Requests: requests, PromptTokens: prompt, CompletionTokens: completion, Cost: cost}, nil
}

func parseQuota(decoder *json.Decoder) (*Quota, error) {
	if decoder == nil {
		return nil, responseInvalid()
	}
	token, err := decoder.Token()
	if err != nil {
		return nil, responseInvalid()
	}
	if token == nil {
		return nil, nil
	}
	if token != json.Delim('{') {
		return nil, responseInvalid()
	}
	fields := map[string]json.Token{}
	seen := map[string]bool{}
	for decoder.More() {
		keyToken, err := decoder.Token()
		key, ok := keyToken.(string)
		if err != nil || !ok || !seenField(seen, key) {
			return nil, responseInvalid()
		}
		value, err := decoder.Token()
		if err != nil {
			return nil, responseInvalid()
		}
		if _, isDelim := value.(json.Delim); isDelim {
			return nil, responseInvalid()
		}
		fields[key] = value
	}
	if token, err := decoder.Token(); err != nil || token != json.Delim('}') {
		return nil, responseInvalid()
	}
	used, err := requiredAmount(fields, "used")
	if err != nil {
		return nil, err
	}
	unitValue, ok := fields["unit"].(string)
	if !ok {
		return nil, responseInvalid()
	}
	unit := QuotaUnit(unitValue)
	if unit != QuotaUSD && unit != QuotaTokens && unit != QuotaRequests {
		return nil, responseInvalid()
	}
	quota := &Quota{Used: used, Unit: unit}
	if limitToken, present := fields["limit"]; present && limitToken != nil {
		amount, err := decodeAmount(limitToken)
		if err != nil {
			return nil, err
		}
		quota.Limit = &amount
	}
	if resets, present := fields["resets_at"]; present && resets != nil {
		value, ok := resets.(string)
		if !ok || !validTime(value) {
			return nil, responseInvalid()
		}
		quota.ResetsAt = value
	}
	return quota, nil
}

func parseModels(decoder *json.Decoder) ([]Model, error) {
	token, err := decoder.Token()
	if err != nil || token != json.Delim('[') {
		return nil, responseInvalid()
	}
	models := make([]Model, 0)
	for decoder.More() {
		fields, err := decodeObject(decoder)
		if err != nil {
			return nil, err
		}
		id, ok := fields["model"].(string)
		if !ok || !validID(id) {
			return nil, responseInvalid()
		}
		requests, err := requiredCount(fields, "requests")
		if err != nil {
			return nil, err
		}
		prompt, err := requiredCount(fields, "prompt_tokens")
		if err != nil {
			return nil, err
		}
		completion, err := requiredCount(fields, "completion_tokens")
		if err != nil {
			return nil, err
		}
		cost, err := requiredAmount(fields, "cost")
		if err != nil {
			return nil, err
		}
		models = append(models, Model{
			Model: id, Requests: requests, PromptTokens: prompt, CompletionTokens: completion, Cost: cost,
		})
		if len(models) > maxModels {
			return nil, responseInvalid()
		}
	}
	if token, err := decoder.Token(); err != nil || token != json.Delim(']') {
		return nil, responseInvalid()
	}
	return models, nil
}

func decodeObject(decoder *json.Decoder) (map[string]json.Token, error) {
	token, err := decoder.Token()
	if err != nil || token != json.Delim('{') {
		return nil, responseInvalid()
	}
	fields := map[string]json.Token{}
	seen := map[string]bool{}
	for decoder.More() {
		keyToken, err := decoder.Token()
		key, ok := keyToken.(string)
		if err != nil || !ok || !seenField(seen, key) {
			return nil, responseInvalid()
		}
		value, err := decoder.Token()
		if err != nil {
			return nil, responseInvalid()
		}
		if _, isDelim := value.(json.Delim); isDelim {
			return nil, responseInvalid()
		}
		fields[key] = value
	}
	if token, err := decoder.Token(); err != nil || token != json.Delim('}') {
		return nil, responseInvalid()
	}
	return fields, nil
}

func seenField(seen map[string]bool, key string) bool {
	if key == "" || sensitiveField(key) || seen[key] {
		return false
	}
	seen[key] = true
	return true
}

func sensitiveField(name string) bool {
	normalized := strings.ToLower(strings.NewReplacer("_", "", "-", "").Replace(name))
	switch normalized {
	case "apikey", "key", "token", "authorization", "secret", "password", "credential", "bearer", "accesstoken", "clientsecret":
		return true
	default:
		return false
	}
}

func decodeString(decoder *json.Decoder) (string, error) {
	token, err := decoder.Token()
	value, ok := token.(string)
	if err != nil || !ok {
		return "", responseInvalid()
	}
	return value, nil
}

func requiredCount(fields map[string]json.Token, key string) (int64, error) {
	value, ok := fields[key]
	if !ok {
		return 0, responseInvalid()
	}
	number, ok := value.(json.Number)
	if !ok {
		return 0, responseInvalid()
	}
	if strings.ContainsAny(number.String(), ".eE") {
		return 0, responseInvalid()
	}
	count, err := number.Int64()
	if err != nil || count < 0 {
		return 0, responseInvalid()
	}
	return count, nil
}

func requiredAmount(fields map[string]json.Token, key string) (float64, error) {
	value, ok := fields[key]
	if !ok {
		return 0, responseInvalid()
	}
	return decodeAmount(value)
}

func decodeAmount(value json.Token) (float64, error) {
	number, ok := value.(json.Number)
	if !ok {
		return 0, responseInvalid()
	}
	amount, err := number.Float64()
	if err != nil || amount < 0 || strings.EqualFold(number.String(), "nan") ||
		strings.EqualFold(number.String(), "inf") || strings.EqualFold(number.String(), "+inf") ||
		strings.EqualFold(number.String(), "-inf") {
		return 0, responseInvalid()
	}
	if _, err := strconv.ParseFloat(number.String(), 64); err != nil {
		return 0, responseInvalid()
	}
	return amount, nil
}

func validTime(value string) bool {
	if value == "" {
		return false
	}
	_, err := time.Parse(time.RFC3339, value)
	return err == nil
}

func validID(id string) bool {
	if id == "" || len(id) > maxIDBytes || !utf8.ValidString(id) {
		return false
	}
	first, _ := utf8.DecodeRuneInString(id)
	last, _ := utf8.DecodeLastRuneInString(id)
	if unicode.IsSpace(first) || unicode.IsSpace(last) {
		return false
	}
	for _, r := range id {
		if unicode.IsControl(r) {
			return false
		}
	}
	return true
}

func skipValue(decoder *json.Decoder) error {
	token, err := decoder.Token()
	if err != nil {
		return err
	}
	delim, ok := token.(json.Delim)
	if !ok {
		return nil
	}
	switch delim {
	case '{':
		for decoder.More() {
			if _, err := decoder.Token(); err != nil {
				return err
			}
			if err := skipValue(decoder); err != nil {
				return err
			}
		}
		_, err = decoder.Token()
		return err
	case '[':
		for decoder.More() {
			if err := skipValue(decoder); err != nil {
				return err
			}
		}
		_, err = decoder.Token()
		return err
	default:
		return responseInvalid()
	}
}

File: desktop/src-tauri/src/manager_core/apikeyusage/parse.rs
use std::collections::{HashMap, HashSet};
use std::fmt;

use serde::de::{self, Deserializer, IgnoredAny, MapAccess, SeqAccess, Visitor};
use serde::Deserialize;
use serde_json::Number;

use super::error::{invalid_period, response_invalid, UsageError};

pub const MAX_MODELS: usize = 64;
pub const MAX_QUOTAS: usize = 32;
const MAX_ID_BYTES: usize = 256;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Period {
    OneHour,
    TwelveHours,
    TwentyFourHours,
    SevenDays,
    ThirtyDays,
    Today,
    ThisWeek,
    ThisMonth,
}

impl Period {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::OneHour => "1h",
            Self::TwelveHours => "12h",
            Self::TwentyFourHours => "24h",
            Self::SevenDays => "7d",
            Self::ThirtyDays => "30d",
            Self::Today => "today",
            Self::ThisWeek => "this_week",
            Self::ThisMonth => "this_month",
        }
    }

    fn parse(value: &str) -> Option<Self> {
        match value {
            "1h" => Some(Self::OneHour),
            "12h" => Some(Self::TwelveHours),
            "24h" => Some(Self::TwentyFourHours),
            "7d" => Some(Self::SevenDays),
            "30d" => Some(Self::ThirtyDays),
            "today" => Some(Self::Today),
            "this_week" => Some(Self::ThisWeek),
            "this_month" => Some(Self::ThisMonth),
            _ => None,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum BudgetPeriod {
    Day,
    Week,
    Month,
}

impl BudgetPeriod {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Day => "day",
            Self::Week => "week",
            Self::Month => "month",
        }
    }

    fn parse(value: &str) -> Option<Self> {
        match value {
            "day" => Some(Self::Day),
            "week" => Some(Self::Week),
            "month" => Some(Self::Month),
            _ => None,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum QuotaUnit {
    Usd,
    Tokens,
    Requests,
}

impl QuotaUnit {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Usd => "usd",
            Self::Tokens => "tokens",
            Self::Requests => "requests",
        }
    }

    fn parse(value: &str) -> Option<Self> {
        match value {
            "usd" => Some(Self::Usd),
            "tokens" => Some(Self::Tokens),
            "requests" => Some(Self::Requests),
            _ => None,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct Snapshot {
    pub period: Period,
    pub as_of: String,
    pub summary: Summary,
    pub quota: Option<Quota>,
    pub quotas: Vec<ProviderQuota>,
    pub by_model: Vec<Model>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Summary {
    pub requests: i64,
    pub prompt_tokens: i64,
    pub completion_tokens: i64,
    pub cost: f64,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Quota {
    pub used: f64,
    pub limit: Option<f64>,
    pub unit: QuotaUnit,
    pub resets_at: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ProviderQuota {
    pub provider: String,
    pub period: BudgetPeriod,
    pub used: f64,
    pub limit: f64,
    pub unit: QuotaUnit,
    pub resets_at: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Model {
    pub model: String,
    pub requests: i64,
    pub prompt_tokens: i64,
    pub completion_tokens: i64,
    pub cost: f64,
}

pub fn normalize_period(value: &str) -> Result<Period, UsageError> {
    match value.trim() {
        "" | "7d" => Ok(Period::SevenDays),
        other => Period::parse(other).ok_or_else(invalid_period),
    }
}

pub fn parse(body: &[u8], period: Period) -> Result<Snapshot, UsageError> {
    if std::str::from_utf8(body).is_err() {
        return Err(response_invalid());
    }
    let parsed: ParsedSnapshot = serde_json::from_slice(body).map_err(|_| response_invalid())?;
    if parsed.period != period {
        return Err(response_invalid());
    }
    Ok(Snapshot {
        period,
        as_of: parsed.as_of,
        summary: parsed.summary,
        quota: parsed.quota,
        quotas: parsed.quotas,
        by_model: parsed.by_model,
    })
}

struct ParsedSnapshot {
    period: Period,
    as_of: String,
    summary: Summary,
    quota: Option<Quota>,
    quotas: Vec<ProviderQuota>,
    by_model: Vec<Model>,
}

impl<'de> Deserialize<'de> for ParsedSnapshot {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        deserializer.deserialize_map(SnapshotVisitor)
    }
}

struct SnapshotVisitor;

impl<'de> Visitor<'de> for SnapshotVisitor {
    type Value = ParsedSnapshot;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("usage object")
    }

    fn visit_map<A: MapAccess<'de>>(self, mut map: A) -> Result<Self::Value, A::Error> {
        let mut seen = HashSet::new();
        let mut period = None;
        let mut as_of = String::new();
        let mut summary = None;
        let mut quota = None;
        let mut quotas = Vec::new();
        let mut by_model = None;
        while let Some(key) = map.next_key::<String>()? {
            if !seen_field(&mut seen, &key) {
                return Err(de::Error::custom("invalid field"));
            }
            match key.as_str() {
                "period" => {
                    let value = map.next_value::<String>()?;
                    period =
                        Some(Period::parse(&value).ok_or_else(|| de::Error::custom("period"))?);
                }
                "as_of" => {
                    let value = map.next_value::<String>()?;
                    if !valid_time(&value) {
                        return Err(de::Error::custom("as_of"));
                    }
                    as_of = value;
                }
                "summary" => summary = Some(map.next_value::<Summary>()?),
                "quota" => quota = map.next_value::<Option<Quota>>()?,
                "quotas" => {
                    quotas = map
                        .next_value::<Option<QuotaRows>>()?
                        .map(|rows| rows.0)
                        .unwrap_or_default();
                }
                "by_model" => by_model = Some(map.next_value::<ModelRows>()?.0),
                _ => {
                    let _: IgnoredAny = map.next_value()?;
                }
            }
        }
        if !seen.contains("period") || !seen.contains("summary") || !seen.contains("by_model") {
            return Err(de::Error::custom("missing required"));
        }
        Ok(ParsedSnapshot {
            period: period.ok_or_else(|| de::Error::custom("period"))?,
            as_of,
            summary: summary.ok_or_else(|| de::Error::custom("summary"))?,
            quota,
            quotas,
            by_model: by_model.ok_or_else(|| de::Error::custom("by_model"))?,
        })
    }
}

impl<'de> Deserialize<'de> for Summary {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let fields = ScalarObject::deserialize(deserializer)?.0;
        Ok(Summary {
            requests: required_count(&fields, "requests")?,
            prompt_tokens: required_count(&fields, "prompt_tokens")?,
            completion_tokens: required_count(&fields, "completion_tokens")?,
            cost: required_amount(&fields, "cost")?,
        })
    }
}

impl<'de> Deserialize<'de> for Quota {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let fields = ScalarObject::deserialize(deserializer)?.0;
        let used = required_amount(&fields, "used")?;
        let unit = match fields.get("unit") {
            Some(Scalar::String(value)) => {
                QuotaUnit::parse(value).ok_or_else(|| de::Error::custom("unit"))?
            }
            _ => return Err(de::Error::custom("unit")),
        };
        let mut quota = Quota {
            used,
            limit: None,
            unit,
            resets_at: String::new(),
        };
        if let Some(limit) = fields.get("limit") {
            if !matches!(limit, Scalar::Null) {
                quota.limit = Some(decode_amount(limit)?);
            }
        }
        if let Some(resets) = fields.get("resets_at") {
            if !matches!(resets, Scalar::Null) {
                let Scalar::String(value) = resets else {
                    return Err(de::Error::custom("resets_at"));
                };
                if !valid_time(value) {
                    return Err(de::Error::custom("resets_at"));
                }
                quota.resets_at = value.clone();
            }
        }
        Ok(quota)
    }
}

impl<'de> Deserialize<'de> for ProviderQuota {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let fields = ScalarObject::deserialize(deserializer)?.0;
        let provider = match fields.get("provider") {
            Some(Scalar::String(value)) if valid_id(value) => value.clone(),
            _ => return Err(de::Error::custom("provider")),
        };
        let period = match fields.get("period") {
            Some(Scalar::String(value)) => {
                BudgetPeriod::parse(value).ok_or_else(|| de::Error::custom("period"))?
            }
            _ => return Err(de::Error::custom("period")),
        };
        let used = required_amount(&fields, "used")?;
        let limit = required_positive_amount(&fields, "limit")?;
        let unit = match fields.get("unit") {
            Some(Scalar::String(value)) if value == "usd" => QuotaUnit::Usd,
            _ => return Err(de::Error::custom("unit")),
        };
        let resets_at = match fields.get("resets_at") {
            Some(Scalar::String(value)) if valid_time(value) => value.clone(),
            _ => return Err(de::Error::custom("resets_at")),
        };
        Ok(ProviderQuota {
            provider,
            period,
            used,
            limit,
            unit,
            resets_at,
        })
    }
}

impl<'de> Deserialize<'de> for Model {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let fields = ScalarObject::deserialize(deserializer)?.0;
        let model = match fields.get("model") {
            Some(Scalar::String(value)) if valid_id(value) => value.clone(),
            _ => return Err(de::Error::custom("model")),
        };
        Ok(Model {
            model,
            requests: required_count(&fields, "requests")?,
            prompt_tokens: required_count(&fields, "prompt_tokens")?,
            completion_tokens: required_count(&fields, "completion_tokens")?,
            cost: required_amount(&fields, "cost")?,
        })
    }
}

struct ModelRows(Vec<Model>);

impl<'de> Deserialize<'de> for ModelRows {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        struct ModelsVisitor;
        impl<'de> Visitor<'de> for ModelsVisitor {
            type Value = ModelRows;
            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("model rows")
            }
            fn visit_seq<A: SeqAccess<'de>>(self, mut seq: A) -> Result<Self::Value, A::Error> {
                let mut models = Vec::new();
                while let Some(model) = seq.next_element::<Model>()? {
                    models.push(model);
                    if models.len() > MAX_MODELS {
                        return Err(de::Error::custom("too many models"));
                    }
                }
                Ok(ModelRows(models))
            }
        }
        deserializer.deserialize_seq(ModelsVisitor)
    }
}

struct QuotaRows(Vec<ProviderQuota>);

impl<'de> Deserialize<'de> for QuotaRows {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        struct QuotasVisitor;
        impl<'de> Visitor<'de> for QuotasVisitor {
            type Value = QuotaRows;
            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("provider quotas")
            }
            fn visit_seq<A: SeqAccess<'de>>(self, mut seq: A) -> Result<Self::Value, A::Error> {
                let mut quotas = Vec::new();
                while let Some(quota) = seq.next_element::<ProviderQuota>()? {
                    quotas.push(quota);
                    if quotas.len() > MAX_QUOTAS {
                        return Err(de::Error::custom("too many quotas"));
                    }
                }
                Ok(QuotaRows(quotas))
            }
        }
        deserializer.deserialize_seq(QuotasVisitor)
    }
}

struct ScalarObject(HashMap<String, Scalar>);

impl<'de> Deserialize<'de> for ScalarObject {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        struct ObjectVisitor;
        impl<'de> Visitor<'de> for ObjectVisitor {
            type Value = ScalarObject;
            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("scalar object")
            }
            fn visit_map<A: MapAccess<'de>>(self, mut map: A) -> Result<Self::Value, A::Error> {
                let mut seen = HashSet::new();
                let mut fields = HashMap::new();
                while let Some(key) = map.next_key::<String>()? {
                    if !seen_field(&mut seen, &key) {
                        return Err(de::Error::custom("invalid field"));
                    }
                    fields.insert(key, map.next_value::<Scalar>()?);
                }
                Ok(ScalarObject(fields))
            }
        }
        deserializer.deserialize_map(ObjectVisitor)
    }
}

enum Scalar {
    Null,
    Bool(bool),
    Number(Number),
    String(String),
}

impl<'de> Deserialize<'de> for Scalar {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        struct ScalarVisitor;
        impl<'de> Visitor<'de> for ScalarVisitor {
            type Value = Scalar;
            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("scalar")
            }
            fn visit_unit<E: de::Error>(self) -> Result<Self::Value, E> {
                Ok(Scalar::Null)
            }
            fn visit_none<E: de::Error>(self) -> Result<Self::Value, E> {
                Ok(Scalar::Null)
            }
            fn visit_bool<E: de::Error>(self, value: bool) -> Result<Self::Value, E> {
                Ok(Scalar::Bool(value))
            }
            fn visit_i64<E: de::Error>(self, value: i64) -> Result<Self::Value, E> {
                Ok(Scalar::Number(value.into()))
            }
            fn visit_u64<E: de::Error>(self, value: u64) -> Result<Self::Value, E> {
                Ok(Scalar::Number(value.into()))
            }
            fn visit_f64<E: de::Error>(self, value: f64) -> Result<Self::Value, E> {
                Number::from_f64(value)
                    .map(Scalar::Number)
                    .ok_or_else(|| de::Error::custom("number"))
            }
            fn visit_str<E: de::Error>(self, value: &str) -> Result<Self::Value, E> {
                Ok(Scalar::String(value.to_owned()))
            }
        }
        deserializer.deserialize_any(ScalarVisitor)
    }
}

fn seen_field(seen: &mut HashSet<String>, key: &str) -> bool {
    if key.is_empty() || sensitive_field(key) || !seen.insert(key.to_owned()) {
        return false;
    }
    true
}

fn sensitive_field(name: &str) -> bool {
    let normalized: String = name
        .chars()
        .filter(|ch| *ch != '_' && *ch != '-')
        .flat_map(|ch| ch.to_lowercase())
        .collect();
    matches!(
        normalized.as_str(),
        "apikey"
            | "key"
            | "token"
            | "authorization"
            | "secret"
            | "password"
            | "credential"
            | "bearer"
            | "accesstoken"
            | "clientsecret"
    )
}

fn required_count<E: de::Error>(fields: &HashMap<String, Scalar>, key: &str) -> Result<i64, E> {
    match fields.get(key) {
        Some(Scalar::Number(number)) => decode_count(number),
        _ => Err(de::Error::custom(key)),
    }
}

fn required_amount<E: de::Error>(fields: &HashMap<String, Scalar>, key: &str) -> Result<f64, E> {
    match fields.get(key) {
        Some(value) => decode_amount(value),
        None => Err(de::Error::custom(key)),
    }
}

fn required_positive_amount<E: de::Error>(
    fields: &HashMap<String, Scalar>,
    key: &str,
) -> Result<f64, E> {
    let amount = required_amount(fields, key)?;
    if amount <= 0.0 {
        return Err(de::Error::custom(key));
    }
    Ok(amount)
}

fn decode_count<E: de::Error>(number: &Number) -> Result<i64, E> {
    let text = number.to_string();
    if text.contains(['.', 'e', 'E']) {
        return Err(de::Error::custom("count"));
    }
    let count = number.as_i64().ok_or_else(|| de::Error::custom("count"))?;
    if count < 0 {
        return Err(de::Error::custom("count"));
    }
    Ok(count)
}

fn decode_amount<E: de::Error>(value: &Scalar) -> Result<f64, E> {
    let Scalar::Number(number) = value else {
        return Err(de::Error::custom("amount"));
    };
    let text = number.to_string();
    if text.eq_ignore_ascii_case("nan")
        || text.eq_ignore_ascii_case("inf")
        || text.eq_ignore_ascii_case("+inf")
        || text.eq_ignore_ascii_case("-inf")
    {
        return Err(de::Error::custom("amount"));
    }
    let amount = number.as_f64().ok_or_else(|| de::Error::custom("amount"))?;
    if amount < 0.0 || !amount.is_finite() {
        return Err(de::Error::custom("amount"));
    }
    Ok(amount)
}

fn valid_time(value: &str) -> bool {
    !value.is_empty() && chrono::DateTime::parse_from_rfc3339(value).is_ok()
}

fn valid_id(id: &str) -> bool {
    if id.is_empty() || id.len() > MAX_ID_BYTES {
        return false;
    }
    let mut chars = id.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    let last = chars.next_back().unwrap_or(first);
    if first.is_whitespace() || last.is_whitespace() {
        return false;
    }
    id.chars().all(|ch| !ch.is_control())
}
